pkg/scheduler/queue/query_component_capacity.go

```diff
@@ -0,0 +1,177 @@
+	// Annotations from the frontend representing "both" or "unknown" do not need to be checked explicitly.
+	// Conservatively, we assume a query request will be served by both query components
+	// as we prefer to overestimate query component capacity utilization rather than underestimate.
+	isIngester, isStoreGateway = true, true
+
+	switch queryComponentName {
+	case ingesterQueueDimension:
+		isStoreGateway = false
+	case storeGatewayQueueDimension:
+		isIngester = false
+	}
+	return isIngester, isStoreGateway
+}
+
+// QueryComponentCapacity tracks requests from the time they are forwarded to a querier
+// to the time are completed by the querier or failed due to cancel, timeout, or disconnect.
+// Unlike the Scheduler's schedulerInflightRequests, tracking begins only when the request is sent to a querier.
+//
+// Scheduler-Querier inflight requests are broken out by the query component flags,
+// representing whether the query request will be served by the ingesters, store-gateways, or both.
+// Query requests utilizing both ingesters and store-gateways are tracked in the atomics for both component,
+// therefore the sum of inflight requests by component is likely to exceed the inflight requests total.
+type QueryComponentCapacity struct {
+	// targetReservedCapacity sets the portion of querier-worker connections we attempt to reserve
+	// for queries to the less-loaded query component when the query queue becomes backlogged.
+	targetReservedCapacity float64
+
+	ingesterInflightRequests     *atomic.Int64
+	storeGatewayInflightRequests *atomic.Int64
+	querierInflightRequestsTotal *atomic.Int64
+
+	querierInflightRequestsGauge *prometheus.GaugeVec
+}
+
+// DefaultReservedQueryComponentCapacity reserves 1 / 3 of querier-worker connections
+// for the query component utilizing fewer of the available connections.
+const DefaultReservedQueryComponentCapacity = 0.33
+
+const MinReservedQueryComponentCapacity = 0.1
+const MaxReservedQueryComponentCapacity = 0.5
+
+func NewQueryComponentCapacity(
+	targetReservedCapacity float64,
+	querierInflightRequests *prometheus.GaugeVec,
+) (*QueryComponentCapacity, error) {
+
+	if targetReservedCapacity < MinReservedQueryComponentCapacity || targetReservedCapacity >= MaxReservedQueryComponentCapacity {
+		return nil, errors.New("invalid targetReservedCapacity")
+	}
+
+	return &QueryComponentCapacity{
+		targetReservedCapacity: targetReservedCapacity,
+
+		ingesterInflightRequests:     atomic.NewInt64(0),
+		storeGatewayInflightRequests: atomic.NewInt64(0),
+		querierInflightRequestsTotal: atomic.NewInt64(0),
+
+		querierInflightRequestsGauge: querierInflightRequests,
+	}, nil
+}
+
+// ExceedsCapacityForComponentName checks whether a query component has exceeded the capacity utilization threshold.
+// This enables the dequeuing algorithm to skip requests for a query component experiencing heavy load,
+// reserving querier-worker connection capacity to continue servicing requests for the other component.
```

Comment:
I also think we need to track three kinds of queries, rather than whether they touch ingesters or store-gateways:
* ingester-only
* store-gateway-only
* ingester and store-gateway

The reason for this is that otherwise, there's a risk that requests that touch both components would starve requests that touch only one. (For example, imagine a scenario where store-gateways are running slowly - all of the capacity we'd like to allocate to ingester-only queries could be consumed by queries that touch both ingesters and store-gateways.)

Reply:
this is will be handled by the queue structure that Casie is working on, using this method. this method gives that exact information - if something touches both and either one has exceeded the threshold it will let you know

Reply:
I think I'm missing something - how does this method track "ingester+store-gateway" requests separately from "ingester-only" requests?

Reply:
It's not this components job to "track" anything about the requests or queue at all.
It is simply going to tell the other component (being worked on by Casie) that navigates through the queue structure whether any of the query components (ingester or store gateway) exceeds the reserved capacity _for that expected component name_, where the expected component name is expected to be `ingester`, `store-gateway`, `ingester-and-store-gateway`, or other (which is treated as ingester and store gateway).

`isIngester, isStoreGateway := queryComponentFlags(name)`
both of those flags will be true if the expected component name is anything other than just "ingester" or just "store-gateway".

When both flags are true, the capacity utliization is checked for both.
that's the `if isIngester { // return if load exceeded }` and `if isStoreGateway { // return if load exceeded }` parts of this method.